pkg/member/participant.go

```diff
@@ -0,0 +1,405 @@
+// Copyright 2016 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/election"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+var _ Member = (*Participant)(nil)
+
+// Participant is used for the election related logic. Compared to its counterpart
+// EmbeddedEtcdMember, Participant relies on etcd for election, but it's decoupled
+// with the embedded etcd. It implements Member interface.
+type Participant struct {
```

Comment:
I don't quite understand the purpose of `Participant` here. What is the difference between member and if it decouples with etcd, why do we still have the concept of the leader?

Reply:
The leader is the concept used by pkg/election for leader election. The elected leader will be the Primary in a primary/standby cluster configuration. They are two different concepts used at different levels in the system.